specs/networking/rpc-interface.md

```diff
@@ -0,0 +1,292 @@
+         |
+       +-+-+
+       |   |  +-----------+ Epoch 1
+       +---+
+```
+
+Once the handshake completes, the client with the higher `latest_finalized_epoch` or `best_slot` (if the clients have equal `latest_finalized_epoch`s) SHOULD request beacon block roots from its counterparty via `beacon_block_roots` (i.e., RPC method `10`).
+
+### Goodbye
+
+**Method ID:** `1`
+
+**Body:**
+
+```
+(
+    reason: uint64
+)
+```
+
+Client MAY send `goodbye` messages upon disconnection. The reason field MAY be one of the following values:
+
+- `1`: Client shut down.
+- `2`: Irrelevant network.
+- `3`: Fault/error.
+
+Clients MAY define custom goodbye reasons as long as the value is larger than `1000`.
+
+### Get Status
+
+**Method ID:** `2`
+
+**Request Body:**
+
+```
+(
+	sha: bytes32
+	user_agent: bytes
+	timestamp: uint64
+)
+```
+
+**Response Body:**
+
+```
+(
+	sha: bytes32
+	user_agent: bytes
+	timestamp: uint64
+)
+```
+
+Returns metadata about the remote node.
+
+### Request Beacon Block Roots
+
+**Method ID:** `10`
+
+**Request Body**
+
+```
+(
+	start_slot: uint64
+	count: uint64
+)
+```
+
+**Response Body:**
+
+```
+# BlockRootSlot
+(
+    block_root: bytes32
+    slot: uint64
+)
+
+(
+    roots: []BlockRootSlot
+)
+```
+
+Requests a list of block roots and slots from the peer. The `count` parameter MUST be less than or equal to `32768`. The slots MUST be returned in ascending slot order.
+
+### Beacon Block Headers
+
+**Method ID:** `11`
+
+**Request Body**
+
+```
+(
+    start_root: HashTreeRoot
+    start_slot: uint64
+    max_headers: uint64
+    skip_slots: uint64
+)
+```
+
+**Response Body:**
+
+```
+(
+    headers: []BeaconBlockHeader
+)
+```
+
+Requests beacon block headers from the peer starting from `(start_root, start_slot)`. The response MUST contain no more than `max_headers` headers. `skip_slots` defines the maximum number of slots to skip between blocks. For example, requesting blocks starting at slots `2` a `skip_slots` value of `1` would return the blocks at `[2, 4, 6, 8, 10]`. In cases where a slot is empty for a given slot number, the closest previous block MUST be returned. For example, if slot `4` were empty in the previous example, the returned array would contain `[2, 3, 6, 8, 10]`. If slot three were further empty, the array would contain `[2, 6, 8, 10]` - i.e., duplicate blocks MUST be collapsed. A `skip_slots` value of `0` returns all blocks.
+
+The function of the `skip_slots` parameter helps facilitate light client sync - for example, in [#459](https://github.com/ethereum/eth2.0-specs/issues/459) - and allows clients to balance the peers from whom they request headers. Clients could, for instance, request every 10th block from a set of peers where each per has a different starting block in order to populate block data.
+
+### Beacon Block Bodies
+
+**Method ID:** `12`
+
+**Request Body:**
+
+```
+(
+    block_roots: []HashTreeRoot
+)
+```
+
+**Response Body:**
+
+```
+(
+    block_bodies: []BeaconBlockBody
```

Comment:
To obtain blocks from a peer I expect to first request headers, then bodies. After receiving the bodies, headers are combined with the bodies and `chain.import_block(..)` (or w/e) is called.

Given that we don't have any information linking `body` to `header`, we need to iterate headers to find `header.block_body_root == hash_tree_root(body)`.

You could reduce this iteration to "all headers from that `peer_id`" rather than _all_ headers, I guess.

Nonetheless, thoughts on returning `(block_body, block_root)` to allow `O(1)` lookup of corresponding header?

Reply:
Oh ignore this, I can just key the headers by `block_body_root`!